### PR TITLE
gradient with infinite elements (2)

### PR DESCRIPTION
**Pull request by BalticPinguin**
This is the corrected version of #1969, having some minor changes (and rebased to current master).

 * 671fe8b  contains some minor fixes in the documentations: The main changes are to name the variables in the declaration (.h-file) and the function (.C-file) consistent.
    The main purpose is, however, to adapt the MeshFunction::gradient()-function to infinite elements.
    Respectively, several functions and objects are added to FEInterface.

I have one open question related to this, denoted by the FIXME-tag: To get the gradient in physical coordinates, I need to hold the transformation in the FEInterface; therefore, I added the std::vector<std::vector<Real>> local_transform variable; but maybe there is a better way to do so?
I would think of some Tensor- or matrix-type or so?

Any other thoughts about this?
I hope, you are ok with the introduction of this new feature?

Best,
Hubert

### CODE REVIEW

File: src/fe/inf_fe.C [hunk 1]
```diff
@@ -217,7 +217,7 @@ void InfFE<Dim,T_radial,T_map>::reinit(const Elem * inf_elem,
       // right now, and it will generalize a bit, and it won't break
       // the assumptions elsewhere in InfFE.
       std::vector<Point> radial_pts;
-      for (auto p : index_range(*pts))
+      for (std::size_t p=1; p < pts->size(); ++p)
```
Comment (roystgnr) [inline src/fe/inf_fe.C]: What happened here?  Why are you skipping p=0?

File: src/mesh/mesh_function.C [hunk 2]
```diff
@@ -556,19 +590,55 @@ void MeshFunction::discontinuous_gradient (const Point & p,
...
+          // Otherwive it is wrong...
+#ifdef LIBMESH_ENABLE_INFINITE_ELEMENTS
+          if (!element->infinite())
+            {
+#endif
+              std::unique_ptr<FEBase> point_fe (FEBase::build(dim, fe_type));
+              const std::vector<std::vector<RealGradient>> & dphi = point_fe->get_dphi();
+              point_fe->reinit(element, & point_list);
+
+              temp_output[index] = grad;
```
Comment (roystgnr) [inline src/mesh/mesh_function.C]: Should this be here?  It looks like it's redundantly zeroing out something we'll write the *correct* value to later.

File: src/mesh/mesh_function.C [hunk 3]
```diff
@@ -597,6 +667,13 @@ void MeshFunction::hessian (const Point & p,
     }
   else
     {
+#ifdef LIBMESH_ENABLE_INFINITE_ELEMENTS
+      if(element->infinite())
+        libmesh_warning("Warning: Requested the Hessian of an Infinit element."
```
Comment (roystgnr) [inline src/mesh/mesh_function.C]: s/Infinit/Infinite

**Comment by roystgnr** [comment]
Those are the only flaws I can find.  :+1: after they're fixed, though I'd feel at least 10 times safer if you could add some unit test coverage for the new features.

File: src/fe/inf_fe.C (hunk 1 above)
Comment (BalticPinguin) [reply src/fe/inf_fe.C]: Oh; you are right. Here I must have mixed up the commit my merging the master...

**Comment by BalticPinguin** [comment]
Yes, I completely agree with you: I think none of the changes currently are tested.
I have still some things in my application code which are not working as expected, but I think the given changes will be incorporated to the miscellaneous example 14.
Many thanks for the review.

File: src/mesh/mesh_function.C (hunk 2 above)
Comment (BalticPinguin) [reply src/mesh/mesh_function.C]: you are right; this is not necessary at all.

**Pull request merged**